### PR TITLE
Fix incorrect parsing of `SIGSTRUCT.DATE`

### PR DESCRIPTION
**Pull request by kailun-qin**
## Description of the changes 
This is a commit corresponding to ["[python] Fix wrong format of `SIGSTRUCT.DATE`"](https://github.com/gramineproject/gramine/pull/1287) commit in core Gramine.

Fixes https://github.com/gramineproject/gsc/issues/159.

## How to test this PR? 

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @kailun-qin)

___
*[`gsc.py` line 419 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5FZ5z6lvtNIn_1Qm9:-N_5FZ6-7rOYIKzoGTHD:b-396j1s) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L419)):*
> ```Python
>     # Field format: (offset, type, value)
>     fields = {
>         'date': (SGX_ARCH_ENCLAVE_CSS_DATE, '<BBH', 'day', 'month', 'year'),
> ```

Can we split this into three fields, similar to how we have it here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L22-L24
___
*[`gsc.py` line 424 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5FfoFEqZMnpMoOaPD:-N_5FfoFEqZMnpMoOaPE:b-7jiqqp) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L424)):*
> ```Python
>         'isv_prod_id': (SGX_ARCH_ENCLAVE_CSS_ISV_PROD_ID, '<H', 'isv_prod_id'),
>         'isv_svn': (SGX_ARCH_ENCLAVE_CSS_ISV_SVN, '<H', 'isv_svn'),
>         'attributes': (SGX_ARCH_ENCLAVE_CSS_ATTRIBUTES, '8s8s', 'flags', 'xfrms'),
> ```

Can we split this into two fields, similar to how we have it here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L32-L33
___
*[`gsc.py` line 437 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5Fp9d0dNInKa_UmgJ:-N_5Fp9d0dNInKa_UmgK:b1ht40p) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L437)):*
> ```Python
>                 except ValueError:
>                     print(f'Misencoded date_{key} in SIGSTRUCT! '
>                           f'Please consider generating a new SIGSTRUCT with "gramine-sgx-sign".',
> ```

This message makes no sense, because `gramine-sgx-sign` invocation is buried deep inside GSC.

I suggest to print `Misencoded date_{key}`, and that's enough. Don't exit the process, just continue. This is a "sigstruct dump" helper anyway, so we can print warnings and continue here.

**Review by kailun-qin — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

___
*[`gsc.py` line 419 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5FZ5z6lvtNIn_1Qm9:-N_7IDy68vWbg5ywDUCE:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L419)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Can we split this into three fields, similar to how we have it here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L22-L24
</blockquote></details>

Done.
___
*[`gsc.py` line 424 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5FfoFEqZMnpMoOaPD:-N_7IK15-iZkwg7yDkev:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L424)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Can we split this into two fields, similar to how we have it here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L32-L33
</blockquote></details>

Done.
___
*[`gsc.py` line 437 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5Fp9d0dNInKa_UmgJ:-N_7ILuTCN4qqoBzpfL1:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L437)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This message makes no sense, because `gramine-sgx-sign` invocation is buried deep inside GSC.

I suggest to print `Misencoded date_{key}`, and that's enough. Don't exit the process, just continue. This is a "sigstruct dump" helper anyway, so we can print warnings and continue here.
</blockquote></details>

Done.

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners

*[a discussion](https://reviewable.io/reviews/gramineproject/gsc/160#-N_80pV22rv5aZqQTCDY:-N_80pV22rv5aZqQTCDZ:b-dxyh6j) (no related file):*
FYI: I verified this PR on a Python workload manually:
```
$ ./gsc info-image gsc-python:bullseye
...
date = "2023-07-12"
```

Looks good!

___
___
*[`gsc.py` line 437 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5Fp9d0dNInKa_UmgJ:-N_7zlR14NBpVnYa3lJy:b1tqo9z) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L437)):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

But you still do a `raise`, which will terminate the whole GSC?

Actually, I don't mind, this should never happen in practice. And if this does happen, then something is very wrong, and it's better to fail loudly.

**Review by kailun-qin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners

___
*[`gsc.py` line 437 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5Fp9d0dNInKa_UmgJ:-N_87WOq0hzKOmvng5o7:b-fcyor9) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L437)):*
> But you still do a `raise`, which will terminate the whole GSC?

Yes.

> this should never happen in practice

This could happen if one uses a newer GSC (w/ this fix) with an older core Gramine (w/o the corresponding fix). But GSC releases (or master) are guaranteed to work with corresponding Gramine releases (or master) only? So I think it's fine that we just fail loudly.

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners

___
*[`gsc.py` line 437 at r1](https://reviewable.io/reviews/gramineproject/gsc/160#-N_5Fp9d0dNInKa_UmgJ:-N_8Heio4ZTtvwCe5IWa:bpv7fs6) ([raw file](https://github.com/gramineproject/gsc/blob/7ab3cec8674c6288dc5ef8f3cc02e2f50f35ba6c/gsc.py#L437)):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

> But you still do a `raise`, which will terminate the whole GSC?

Yes.

> this should never happen in practice

This could happen if one uses a newer GSC (w/ this fix) with an older core Gramine (w/o the corresponding fix). But GSC releases (or master) are guaranteed to work with corresponding Gramine releases (or master) only? So I think it's fine that we just fail loudly.
</blockquote></details>

Yes, GSC and Gramine should track each other's versions. So we're good.

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, 3 unresolved discussions, not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin)

___
*[`gsc.py` line 434 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aCB2S0am2laNfzLc4:-N_aCB2S0am2laNfzLc5:bi8fr5o) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L434)):*
> ```Python
>         if key in ['date_year', 'date_month', 'date_day']:
>             try:
>                 attr[key] = int(f'{struct.unpack_from(fmt, sig, offset)[0]:x}')
> ```

Umm, what's this? :scream:



_Code quote:_
```Python
int(f'{struct.unpack_from(fmt, sig, offset)[0]:x}')
```
___
*[`gsc.py` line 435 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aCa576G9lbD0tH9Dq:-N_aCa576G9lbD0tH9Dr:b-pmn2ao) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L435)):*
> ```Python
>             try:
>                 attr[key] = int(f'{struct.unpack_from(fmt, sig, offset)[0]:x}')
>             except ValueError:
> ```

How can this happen?
___
*[`gsc.py` line 473 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aBeu2BehVf2rqsbH4:-N_aBeu2BehVf2rqsbH5:bfutn6j) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L473)):*
> ```Python
>             sigstruct['isv_prod_id'] = attr['isv_prod_id']
>             sigstruct['isv_svn'] = attr['isv_svn']
>             sigstruct['date'] = '%d-%02d-%02d' % (attr['date_year'], attr['date_month'], attr['date_day'])
> ```

please wrap

**Review by dimakuv — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, 3 unresolved discussions, not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @kailun-qin and @mkow)

___
*[`gsc.py` line 434 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aCB2S0am2laNfzLc4:-N_crdE1BYssYJdz_dao:bu2t4ad) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L434)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Umm, what's this? :scream:
</blockquote></details>

Directly copied from here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L149

There's a reason why it looks so scary: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L106-L114
___
*[`gsc.py` line 435 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aCa576G9lbD0tH9Dq:-N_ct3vRCdm_HO8zjbzs:b2fb8uu) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L435)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

How can this happen?
</blockquote></details>

Basically, we read the value from sigstruct as if it's a hex string, then convert this hex string to int. If the hex string cannot be converted, then we'll get a ValueError.

**Review by kailun-qin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: 0 of 1 files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow)

___
*[`gsc.py` line 473 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aBeu2BehVf2rqsbH4:-N_d2CRw7jeJs2jPTXCc:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L473)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

please wrap
</blockquote></details>

Done.

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

___
*[`gsc.py` line 434 at r2](https://reviewable.io/reviews/gramineproject/gsc/160#-N_aCB2S0am2laNfzLc4:-NaD81hh2GnUIVDcNt1X:bb7wf3a) ([raw file](https://github.com/gramineproject/gsc/blob/a3441fbdd554140f81b86847c9023d8e15a7b3f8/gsc.py#L434)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Directly copied from here: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L149

There's a reason why it looks so scary: https://github.com/gramineproject/gramine/blob/db2a92e4e37acd7e5023cd728fb78df00527d14d/python/graminelibos/sigstruct.py#L106-L114
</blockquote></details>

Oh, I already forgot how bad it is.

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/160)* status: :shipit: complete! all files reviewed, all discussions resolved
